packages/flutter/lib/src/rendering/proxy_box.dart

```diff
@@ -845,6 +845,16 @@ class RenderOpacity extends RenderProxyBox {
   @override
   bool get alwaysNeedsCompositing => child != null && (_alpha > 0);
```

Comment:
The docs for `alwaysNeedsCompositing` need updating. They currently advice to set this to true when your paint function always creates a layer. As this RO demonstrates, it can also be true when your paint function doesn't create a layer (that's now done in updateCompositedLayer).

Reply:
Actually, if you are already a repaintBoudnary, do we still need to set alwaysNeedsCompositing? Or does that only need to be set if you push additional layers in your paint method (not counting the one created by updateCompositeLayer)? My feeling is that RenderOpacity probably doesn;t need to set alwaysNeedsCompositing anymore because it doesn't push extra layers...

Reply:
Either way, somewhere we need to document the relationship between alwaysNeedsCompositing and isRepaintBoundary/updateCompositeLayer.

Reply:
Good points, I will investigate

Reply:
This comment is also still open.

Reply:
So from more investigation `isRepaintBoundary` is basically a superset of `alwaysNeedsCompositing`, that is the internal compositing bit is tracked with isRepaintBoundary || alwaysNeedsCompositing.

I think we could document that for repaintBOundaries it should remain true, but we don't actually have to care about its state necessarily

Reply:
This sounds like RenderOpacity should no longer override alwaysNeedsCompositing and only control that behavior via the isRepaintBoundary override to simplify things.

(same for the other class below)

Reply:
done

Reply:
When I looked at this a while ago I got the impression that "alwaysNeedsCompositing" was really "I need compositing regardless of my children" and "needsCompositing" was "either myself or my children need compositing".

Many ROs seemed to set the always flag when they had a child even though they could accomplish what the needed to do without inserting a layer.

RepaintBoundary widgets should be the main use of the always flag. Layers like Opacity that could render their modification with a saveLayer if their children don't need compositing shouldn't be forcing compositing (but that one does).

Reply:
But sometimes layers like opacity can't render their children without  pushing a layer, in that case they might as well be repaint boundaries. In some cases like animated opacity, we should be able to animate the opacity while keeping the picture stable.

This change doesn't conflict with the proposal to composite less frequently.

Reply:
> But sometimes layers like opacity can't render their children without pushing a layer, in that case they might as well be repaint boundaries. In some cases like animated opacity, we should be able to animate the opacity while keeping the picture stable.

But right now, these layers cannot ever paint themselves without pushing a layer. Just because they might not isn't a reason they shouldn't try.

> This change doesn't conflict with the proposal to composite less frequently.

Indeed. I just want to keep in mind that compositing is never the *only* option while we revisit the painting behavior of these widgets/ROs.